android/sdl_android/src/main/java/com/smartdevicelink/proxy/SdlProxyBase.java

```diff
@@ -2093,7 +2100,66 @@ private void handleErrorsFromInternalMessageDispatcher(String info, Exception e)
 		_proxyListener.onError("Proxy callback dispatcher is down. Proxy instance is invalid.", e);
 	}
 	/************* END Functions used by the Message Dispatching Queues ****************/
-	
+
+	private OnRPCNotificationListener mEncryptionReqListener = new OnRPCNotificationListener() {
+		@Override
+		public void onNotified(RPCNotification notification) {
+			List<PermissionItem> permissionItems = ((OnPermissionsChange)notification).getPermissionItem();
+			Set<String> encryptedRpcs = new HashSet<>();
+			boolean requireEncryptionAppLevel = Boolean.TRUE.equals(((OnPermissionsChange) notification).getEncryptionRequirement());
+			for (PermissionItem permissionItem : permissionItems) {
+				if (permissionItems != null && !permissionItems.isEmpty()) {
+					if (permissionItem != null) {
+					    if (requireEncryptionAppLevel && Boolean.TRUE.equals(permissionItem.getEncryptionRequirement())) {
+							String rpcName = permissionItem.getRpcName();
+							if (rpcName != null) {
+								encryptedRpcs.add(rpcName);
+							}
+						}
+					}
+				}
+			}
+			if (!encryptedRpcs.isEmpty()) {
+				if (!mRPCSecuredServiceStarted) {
+					startProtectedRPCService();
+				}
+				if (!mEncryptedRPCNames.equals(encryptedRpcs)) {
+					mEncryptedRPCNames = encryptedRpcs;
+				}
+			}
+		}
+	};
+
+	private ISdlServiceListener mRPCSecuredServiceListener = new ISdlServiceListener() {
+		@Override
+		public void onServiceStarted(SdlSession session, SessionType type, boolean isEncrypted) {
+			if(SessionType.RPC.equals(type) && session != null ){
+				Log.v(TAG, "onServiceStarted, sessionID: " + session.getSessionId() + ", session Type: " + type.getName()
+						+ ", isEncrypted: " + isEncrypted);
+				mRPCSecuredServiceStarted = isEncrypted;
+			}
+		}
+
+		@Override
+		public void onServiceEnded(SdlSession session, SessionType type) {
+			if (session != null) {
+				Log.v(TAG, "onServiceEnded, sessionID: " + session.getSessionId() + ", session Type: " + type.getName());

```

Comment:
If the session has ended and the type is RPC, it should be safe to set mRPCSecuredServiceStarted back to false.

Reply:
Thanks Markos, reset `mRPCSecuredServiceStarted` where applicable.